shared/src/balancer_event_handler.rs

```diff
@@ -7,31 +7,107 @@ use crate::{
 use anyhow::{anyhow, Context, Result};
 use contracts::{
     balancer_v2_vault::{
-        self, event_data::PoolRegistered as ContractPoolRegistered, Event as ContractEvent,
+        self,
+        event_data::{
+            PoolRegistered as ContractPoolRegistered, TokensRegistered as ContractTokensRegistered,
+        },
+        Event as ContractEvent,
     },
     BalancerV2Vault,
 };
 use ethcontract::common::DeploymentInformation;
 use ethcontract::{dyns::DynWeb3, Event as EthContractEvent, EventMetadata, H160, H256};
-use std::fmt::Debug;
-use std::ops::RangeInclusive;
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::Debug,
+    ops::RangeInclusive,
+};
 use tokio::sync::Mutex;
 
-#[derive(Debug)]
+#[derive(Clone, Debug)]
 pub enum BalancerEvent {
     PoolRegistered(PoolRegistered),
+    TokensRegistered(TokensRegistered),
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub struct PoolRegistered {
     pub pool_id: H256,
     pub pool_address: H160,
     pub specialization: PoolSpecialization,
 }
 
+#[derive(Debug, Clone)]
+pub struct TokensRegistered {
+    pub pool_id: H256,
+    pub tokens: Vec<H160>,
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq, Hash)]
+pub struct WeightedPool {
+    pool_id: H256,
+    pool_address: Option<H160>,
+    tokens: Option<Vec<H160>>,
+    specialization: Option<PoolSpecialization>,
```

Comment:
The fact that these are never really None is a pit hacky. Maybe we can have a private builder struct or the like but have the publicly facing struct only contain non-nullable values?

Reply:
I found this article on rust builder patterns. Will look at it tomorrow and make appropriate changes!

https://doc.rust-lang.org/1.0.0/style/ownership/builders.html

Reply:
I have taken the route of renaming the Optional `WeightedPool` to `WeightedPoolBuilder` and implementing `TryInto<WeightedPool> for WeightedPoolBuilder` I hope this is what you were going for.